uhyve-interface/src/lib.rs

```diff
@@ -0,0 +1,112 @@
+
+#![no_std]
+
+// TODO: Throw this out, once https://github.com/rust-lang/rfcs/issues/2783 is resolved
+use num_enum::TryFromPrimitive;
+
+pub mod elf;
+pub mod parameters;
+use parameters::*;
+
+/// The version of the uhyve interface. Note: This is not the same as the semver of the crate but
+/// should be increased on every version bump.
+pub const UHYVE_INTERFACE_VERSION: u32 = 1;
+
+/// Enum containing all valid port mappings for hypercalls.
+///
+/// The discriminants of this enum are the respective ports, so one can get the code by calling
+/// e.g., `HypercallPorts::FileWrite as u16`.
+#[non_exhaustive]
+#[repr(u16)]
+#[derive(Debug, Eq, PartialEq, TryFromPrimitive)]
+pub enum HypercallAddress {
+	/// Port address = `0x400`
+	FileWrite = 0x400,
+	/// Port address = `0x440`
+	FileOpen = 0x440,
+	/// Port address = `0x480`
+	FileClose = 0x480,
+	/// Port address = `0x500`
+	FileRead = 0x500,
+	/// Port address = `0x540`
+	Exit = 0x540,
+	/// Port address = `0x580`
+	FileLseek = 0x580,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x640`
+	Netwrite = 0x640,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x680`
+	Netread = 0x680,
+	#[deprecated = "was never really in use"]
+	/// Port address = `0x700`
+	Netstat = 0x700,
+	/// Port address = `0x740`
+	Cmdsize = 0x740,
+	/// Port address = `0x780`
+	Cmdval = 0x780,
+	/// Port address = `0x800`
+	Uart = 0x800,
+	/// Port address = `0x840`
+	FileUnlink = 0x840,
+}
+impl From<Hypercall<'_>> for HypercallAddress {
+	fn from(value: Hypercall) -> Self {
+		match value {
+			Hypercall::Cmdsize(_) => Self::Cmdsize,
+			Hypercall::Cmdval(_) => Self::Cmdval,
+			Hypercall::Exit(_) => Self::Exit,
+			Hypercall::FileClose(_) => Self::FileClose,
+			Hypercall::FileLseek(_) => Self::FileLseek,
+			Hypercall::FileOpen(_) => Self::FileOpen,
+			Hypercall::FileRead(_) => Self::FileRead,
+			Hypercall::FileWrite(_) => Self::FileWrite,
+			Hypercall::FileUnlink(_) => Self::FileUnlink,
+			Hypercall::SerialWriteByte(_) => Self::Uart,
+		}
+	}
+}
+
+/// Hypervisor calls available in uhyve with their respective parameters. See the [module level documentation](crate) on how to invoke them.
+#[non_exhaustive]
+#[derive(Debug)]
+pub enum Hypercall<'a> {
+	/// Get the size of the argument and environment strings. Used to allocate memory for
+	/// [`Hypercall::Cmdval`].
+	Cmdsize(&'a mut CmdsizeParams),
+	/// Copy the argument and environment strings into the VM. Usually preceeeded by
+	/// [`Hypercall::Cmdsize`] so that the guest can allocate the memory for this call.
+	Cmdval(&'a CmdvalParams),
+	/// Exit the VM and return a status.
+	Exit(&'a ExitParams),
+	FileClose(&'a mut CloseParams),
+	FileLseek(&'a mut LseekParams),
+	FileOpen(&'a mut OpenParams),
+	FileRead(&'a mut ReadPrams),
+	FileWrite(&'a WriteParams),
+	FileUnlink(&'a mut UnlinkParams),
```

Comment:
Short docs would be nice. :)

Reply:
True. I was thinking if these calls should be reworked anyway. But I'll add the docs